src/item.cpp

```diff
@@ -771,7 +771,12 @@ bool item::is_frozen_liquid() const
 
 bool item::covers( const bodypart_id &bp ) const
 {
-    return get_covered_body_parts().test( bp.id() );
+    bool does_cover = false;
+    const bodypart_str_id bpid = bp.id();
+    iterate_covered_body_parts_internal( get_side(), [&]( const bodypart_str_id & covered ) {
```

Comment:
Another minor thing:  it's better to use `int_id` (i.e. `bodypart_id` ) as an argument for the callback.  `string_ids` should be used only in two cases:
* when `int_ids` are not available yet (json is not loaded)
* as a static constants

`item::covers` is called when `int_ids` are already available so you can just pass them around to avoid `bp.id()` conversion.  Also note, that `int_id` comparison is (currently) faster than `string_id` comparison.

Reply:
TIL.

Reply:
So, actually, the process of converting `string_id` to `int_id` on the fly is not cheap. It's enough not cheap that it nearly completely negates the efficiency improvements of comparing `int_id` to `int_id`.

Here is the profiling cost of item::covers with int_id comparisons. Basically the way I measure it is by relative to against `get_side()` which is fixed in each run (because I'm not changing it).

![image](https://user-images.githubusercontent.com/667719/97066546-3cc29080-156b-11eb-9ade-497a54c47bc9.png)

Of that cost, 3/4 of the cost of the callback function is `int_id` constructor.

![image](https://user-images.githubusercontent.com/667719/97066571-5a8ff580-156b-11eb-8df4-e4d7f3236866.png)

With `string_id` comparisons, the cost breakdown is thus. *almost* equivalent percentage wise, maybe off by 1% relatively.

![image](https://user-images.githubusercontent.com/667719/97066601-84491c80-156b-11eb-9327-fe454f72d763.png)

And of that, the bulk of the cost is vanilla string comparisons, but a surprising amount is the body of `string_id::operator==` because of the mandatory checks for whether to use the cached int id.

![image](https://user-images.githubusercontent.com/667719/97066614-9b880a00-156b-11eb-94b8-e5cc3eb0bdf0.png)

String equality checks / comparisons are actually astonishingly cheap if you have heterogenous strings, with varying lengths and first characters. They're only expensive if the majority of the time your strings have long common prefixes and identical lengths.

Reply:
That's very unexpected to me. I must ask, are you profiling an optimized build?

If you look at `string_id` -> `int_id` conversion, for static string_ids it's cached (or supposed to be), so it's basically just wrapping a simple `int` in the `int_id` class (hidden behind some function calls).

If you're using optimized build, I'm not sure how this could be slower than string comparison. Maybe, in this specific case both string_ids have cached _cid inside them and valid version so equality always follows the fast path? 
I'd appreciate it if you can dig deeper and find out whether that's the case.

Reply:
Ah, you're right, in case of body part ids strings are short cheap to compare (no indirection, result is known after first chars). 
Thank you for the information!